drivers/place/booking_check_in_helper.cr

```diff
@@ -0,0 +1,279 @@
+    self[:people_present] = @people_present = state
+    apply_state_changes
+  end
+
+  protected def cleanup_state
+    logger.debug { "cleaning up state, pending: #{@meeting_pending}, meeting: #{!!@current_meeting}" }
+    schedule.clear
+
+    meeting = current_meeting
+    if meeting.nil? || (meeting.try(&.id) != @prompted)
+      self[:prompted] = false
+      self[:no_show] = false
+      self[:checked_in] = false
+      self[:responded] = false
+    end
+  end
+
+  protected def apply_state_changes
+    meeting = self.current_meeting
+    return unless meeting && meeting_pending?
+
+    logger.debug { "applying state changes" }
+    schedule.clear
+
+    time_now = Time.utc
+    start_time = meeting.event_start
+    prompt_at = start_time + prompt_after
+    check_presence_from = start_time + present_from
+
+    # Can we auto check-in?
+    if people_present?
+      if time_now >= check_presence_from
+        bookings.start_meeting(start_time.to_unix)
+        schedule.clear
+      else
+        # Schedule an auto check-in check as people_present? might remain high
+        schedule.at(check_presence_from) do
+          bookings.start_meeting(start_time.to_unix)
+          schedule.clear
+        end
+      end
+      return
+    end
+
+    # should we be scheduling a prompt email?
+    if time_now >= prompt_at
+      prompt_user meeting
+    else
+      schedule.at(prompt_at) { prompt_user meeting }
+    end
+  end
+
+  # sends the templated email to the host for booking confirmation
+  protected def prompt_user(meeting : PlaceCalendar::Event)
+    if @prompted == meeting.id
+      logger.debug { "user has already been prompted" }
+      return
+    end
+
+    logger.debug { "prompting user about meeting room booking #{meeting.id}" }
+    params = generate_guest_jwt
+    mailer.send_template(params[:host_email], {"booking", "check_in_prompt"}, params)
+
+    @prompted = meeting.id.not_nil!
+    self[:no_show] = false
+    self[:checked_in] = false
+    self[:responded] = false
+    self[:prompted] = true
+  end
+
+  # processes the response if the user clicks one of the links in the email
+  protected def prompt_response(id : String, check_in : Bool)
+    meeting = current_meeting
+    unless meeting
+      logger.warn { "received response but no current meeting" }
+      return
+    end
+
+    if @prompted == meeting.id && @prompted == id
+      self[:responded] = true
+      logger.info { "host has responded with #{check_in}" }
+
+      if check_in
+        bookings.start_meeting(meeting.event_start.to_unix)
+        self[:checked_in] = true
+      else
+        bookings.end_meeting(meeting.event_start.to_unix)
+        self[:no_show] = true
+      end
+    else
+      logger.warn { "received response for another meeting #{id} != #{meeting.id} or #{@prompted}" }
+    end
+  end
+
+  # generates the parameters that can be mixed into the template email
+  protected def generate_guest_jwt
```

Comment:
This seems to be used across a few driver now. Could it be lifted into an external module?